plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt

```diff
@@ -0,0 +1,355 @@
+import software.aws.toolkits.jetbrains.services.codewhisperer.language.languages.CodeWhispererUnknownLanguage
+import software.aws.toolkits.jetbrains.services.codewhisperer.language.programmingLanguage
+import software.aws.toolkits.jetbrains.services.codewhisperer.settings.CodeWhispererSettings
+import software.aws.toolkits.jetbrains.services.cwc.controller.chat.telemetry.TelemetryHelper
+import software.aws.toolkits.jetbrains.services.cwc.controller.chat.telemetry.getStartUrl
+import java.io.OutputStreamWriter
+import java.net.HttpURLConnection
+import java.net.URL
+import java.util.Stack
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+class ProjectContextProvider(val project: Project, private val encoderServer: EncoderServer) : Disposable {
+    private val scope = disposableCoroutineScope(this)
+
+    private val retryCount = AtomicInteger(0)
+    val isIndexComplete = AtomicBoolean(false)
+    private val mapper = jacksonObjectMapper()
+    init {
+        scope.launch {
+            if (CodeWhispererSettings.getInstance().isProjectContextEnabled()) {
+                while (true) {
+                    if (encoderServer.isNodeProcessRunning()) {
+                        // TODO: need better solution for this
+                        delay(10000)
+                        initAndIndex()
+                        break
+                    } else {
+                        yield()
+                    }
+                }
+            }
+        }
+    }
+    data class IndexRequestPayload(
+        val filePaths: List<String>,
+        val projectRoot: String,
+        val refresh: Boolean
+    )
+
+    data class FileCollectionResult(
+        val files: List<String>,
+        val fileSize: Int
+    )
+
+    data class QueryRequestPayload(
+        val query: String
+    )
+
+    data class UpdateIndexRequestPayload(
+        val filePath: String
+    )
+
+    data class Usage(
+        @JsonIgnoreProperties(ignoreUnknown = true)
+        @JsonProperty("memoryUsage")
+        val memoryUsage: Int? = null,
+        @JsonProperty("cpuUsage")
+        val cpuUsage: Int? = null
+    )
+
+    data class Chunk(
+        @JsonIgnoreProperties(ignoreUnknown = true)
+        @JsonProperty("filePath")
+        val filePath: String? = null,
+        @JsonProperty("content")
+        val content: String? = null,
+        @JsonProperty("id")
+        val id: String? = null,
+        @JsonProperty("index")
+        val index: String? = null,
+        @JsonProperty("vec")
+        val vec: List<String>? = null,
+        @JsonProperty("context")
+        val context: String? = null,
+        @JsonProperty("prev")
+        val prev: String? = null,
+        @JsonProperty("next")
+        val next: String? = null,
+        @JsonProperty("relativePath")
+        val relativePath: String? = null,
+        @JsonProperty("programmingLanguage")
+        val programmingLanguage: String? = null,
+    )
```

Comment:
i dont think these data classes need default values

Reply:
are these really all nullable?

Reply:
is this a blocker?

Reply:
follow up